triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -119,13 +116,22 @@ object Server extends StrictLogging {
                 redirectUri = toRedirectUri(request.uri),
                 clientId = config.clientId,
                 clientSecret = config.clientSecret)
-              val req = HttpRequest(
-                uri = config.oauthUri.withPath(Uri.Path./("token")),
-                entity = HttpEntity(MediaTypes.`application/json`, body.toJson.compactPrint),
-                method = HttpMethods.POST)
+              import com.daml.oauth.server.Request.Token.marshalRequestEntity
               val tokenRequest = for {
+                entity <- Marshal(body).to[RequestEntity]
+                req = HttpRequest(
+                  uri = config.oauthToken,
+                  entity = entity,
+                  method = HttpMethods.POST)
                 resp <- Http().singleRequest(req)
-                tokenResp <- Unmarshal(resp).to[OAuthResponse.Token]
+                tokenResp <- if (resp.status != StatusCodes.OK) {
+                  Unmarshal(resp).to[String].flatMap { msg =>
```

Comment:
I still feel like there must be some builtin pattern for checking the status code and then unmarshalling but I never found one :confused: